File: src_py/Chapter4/computer_firm.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
ComputerFirm模块 - ComputerFirm类的Python实现
转换自Java版本的ComputerFirm.java
"""

import math

"""
@author Gianluca Capone & Davide Sgobba
Python转换

此类包含定义计算机市场中公司异质性的所有变量以及操作这些变量的方法。计算机公司可以是垂直整合的或专业化的。
"""
class ComputerFirm:
    
    def __init__(self, id, pc, start_share, spillover, mod_sys, computer_market):
        """
        构造函数
        
        Args:
            id: 公司标识符
            pc: 如果是PC公司则为True，否则为False
            start_share: 初始市场份额
            spillover: 垂直整合公司中系统研发向组件研发的溢出
            mod_sys: 系统元素的初始设计优点值
            computer_market: 计算机市场对象引用
        """
        # 变量
        self.born = True              # 公司第一次出现时置为"True"，当首次选择供应商后设为"False"
        self.component_rd = 0.0       # 投资于组件研发的资源量 (B-CO_f,t)
        self.contract_d = 0           # 合同持续时间 (T-CO_f,t)
        self.contract_time = 0        # 签订合同的时间 (t_f)
        self.exit_share = start_share # 移动平均市场份额 (SE_f,t)
        self.int_time = 0             # 公司处于整合状态的周期数 (t-I_f)
        self.integrated = False       # 如果公司是垂直整合的则为"True"，否则为"False"
        self.num_of_draws_cmp = 0     # 公司绘制的新潜在组件mod数量 (N-CO_f,t)
        self.num_of_draws_sys = 0     # 公司绘制的新潜在系统mod数量 (N-SY_f,t)
        self.price = 0.0              # 计算机公司收取的单位价格 (P_f,t)
        self.prob_to_int = 0.0        # 公司整合的概率 (Pr(I_f,t))
        self.prob_to_spec = 0.0       # 公司专业化的概率 (Pr(S_f,t))
        self.profit = 0.0             # 公司赚取的利润量 (PI_f,t)
        self.prop_to_int = 0.0        # 公司整合的倾向 (PI_f,t)
        self.prop_to_spec = 0.0       # 公司专业化的倾向 (PS_f,t)
        self.q_sold = 0.0             # 销售的计算机数量 (q_f,t)
        self.share = start_share      # 公司的市场份额 (s_f,t)
        self.supplier_id = -1         # 公司供应商的标识符 (f_f,t)
        self.system_rd = 0.0          # 投资于系统研发的资源量 (B-SY_f,t)
        
        # 技术变量和对象
        self.id = id                  # 公司标识符
        self.t_id = 0                 # 计算机公司使用的组件技术标识符
        self.alive = True             # 如果公司活跃于市场则为"True"，否则为"False"
        self.pc = pc                  # 如果是PC公司则为"True"，否则为"False"
        self.spillover = spillover    # 垂直整合公司中系统研发向组件研发的溢出
        
        # 关联对象
        from .computer import Computer
        from .component import Component
        from .system_element import SystemElement
        
        self.computer = Computer()    # 公司生产的计算机
        self.component = Component(0, self)  # 对于整合公司，公司生产的组件
        self.system = SystemElement(mod_sys, self)  # 公司生产的系统元素
        self.computer_market = computer_market       # 访问计算机市场
        
    def rd_expenditure(self):
        """
        计算研发支出的方法，包括用于系统和组件研发的资源分配
        """
        # 计算总研发支出
        rd = self.computer_market.rd_on_prof * self.profit
        
        if self.integrated:
            # 方程12，垂直整合公司在系统和组件研发之间分配资源
            self.system_rd = rd / (1 + self.spillover)
            self.component_rd = rd - self.system_rd
        else:
            # 非集成企业，所有研发资源分配给系统开发
            self.system_rd = rd
            self.component_rd = 0.0
    
    def progress(self):
        """
        确定公司创新活动的结果
        """
        # 计算可能的创新次数 (方程13)
        temp_num_of_draws = self.system_rd / self.computer_market.draw_cost_sys
        self.num_of_draws_sys = int(temp_num_of_draws)
        remain = temp_num_of_draws - self.num_of_draws_sys
        
        # 处理剩余部分
        random_number = self.computer_market.rng.random()
        if random_number <= remain:
            self.num_of_draws_sys += 1
            
        # 如果公司是垂直整合的，也要计算组件创新
        if self.integrated:
            temp_num_of_draws = self.component_rd / self.computer_market.draw_cost_cmp
            self.num_of_draws_cmp = int(temp_num_of_draws)
            remain = temp_num_of_draws - self.num_of_draws_cmp
            
            random_number = self.computer_market.rng.random()
            if random_number <= remain:
                self.num_of_draws_cmp += 1
        
        # 更新系统元素的mod
        # 计算当前的mu_prog (方程14.a)
        self.system.mu_prog = ((1 - self.computer_market.internal_cum) * 
                             self.computer_market.pk_sys + 
                             self.computer_market.internal_cum * self.system.mod)
        
        z_max = 0.0
        for i in range(1, self.num_of_draws_sys + 1):
            # 从正态分布中抽取
            z = self.system.mu_prog + self.computer_market.sd_sys * self.computer_market.rng.nextGaussian()
            if z > z_max:
                z_max = z
                
        if z_max > self.system.mod:
            self.system.mod = z_max
            
        # 如果公司是垂直整合的，还要更新组件mod
        if self.integrated:
            # 计算当前的组件mu_prog (方程14.b)
            self.component.mu_prog = ((1 - self.computer_market.internal_cum) * 
                                   self.computer_market.pk_cmp[self.t_id] + 
                                   self.computer_market.internal_cum * self.component.mod)
            
            z_max = 0.0
            for i in range(1, self.num_of_draws_cmp + 1):
                # 从正态分布中抽取
                z = self.component.mu_prog + self.computer_market.sd_cmp * self.computer_market.rng.nextGaussian()
                if z > z_max:
                    z_max = z
                    
            if z_max > self.component.mod:
                self.component.mod = z_max
                
            # 更新组件生产成本 (方程11)
            if self.component.mod > 0:
                self.component.production_cost = self.computer_market.nu_cmp / self.component.mod
                
        # 计算整合和专业化的概率
        if self.integrated:
            # 方程16
            tech_age = self.computer_market.t_id_cmp - self.t_id
            
            if tech_age > 0:
                # 计算专业化倾向 (方程16)
                self.prop_to_spec = (self.computer_market.chi0 * 
                                    (tech_age ** self.computer_market.chi1) * 
                                    (self.share ** self.computer_market.chi2))
                
                # 方程17.b
                self.prob_to_spec = 1 - math.exp(-self.computer_market.xi_spec * self.prop_to_spec)
            else:
                self.prop_to_spec = 0.0
                self.prob_to_spec = 0.0
        else:
            # 计算整合倾向 (方程16)
            self.prop_to_int = self.share ** self.computer_market.chi2
            
            # 方程17.a
            self.prob_to_int = 1 - math.exp(-self.computer_market.xi_int * self.prop_to_int)
            
        # 更新计算机产品特性和价格
        if self.integrated:
            # 方程8，9，10
            self.calculate_computer_features_integrated()
        else:
            # 供应商组件特性影响计算机特性
            self.calculate_computer_features_specialized()
            
        # 计算价格 (方程7)
        self.price = self.computer.production_cost * (1 + self.computer_market.markup)
        
    def calculate_computer_features_integrated(self):
        """
        为垂直整合的公司计算计算机特性
        """
        import math
        # 计算性能和低成本性 (方程8，9)
        self.computer.perf = self.system.mod ** (1 - self.computer_market.tau) * self.component.mod ** self.computer_market.tau
        self.computer.cheap = 1 / (self.component.production_cost ** self.computer_market.tau)
        
        # 计算计算机mod (方程10)
        theta_radians = self.computer_market.theta * math.pi / 180.0
        self.computer.mod = self.computer_market.phi * (self.computer.perf * math.cos(theta_radians) + 
                                                     self.computer.cheap * math.sin(theta_radians)) ** self.computer_market.rho
        
        # 计算计算机生产成本 (方程11)
        self.computer.production_cost = self.computer_market.nu_computer / self.computer.cheap
        
    def calculate_computer_features_specialized(self):
        """
        为专业化公司计算计算机特性
        """
        import math
        # 如果有供应商
        if self.supplier_id != -1:
            # 获取供应商组件的mod和生产成本
            component_firm = self.computer_market.model.component.firm[self.supplier_id]
            supplier_mod = component_firm.component.mod
            supplier_cost = component_firm.component.production_cost
            
            # 计算性能和低成本性 (方程8，9)
            self.computer.perf = self.system.mod ** (1 - self.computer_market.tau) * supplier_mod ** self.computer_market.tau
            self.computer.cheap = 1 / (supplier_cost ** self.computer_market.tau)
            
            # 计算计算机mod (方程10)
            theta_radians = self.computer_market.theta * math.pi / 180.0
            self.computer.mod = self.computer_market.phi * (self.computer.perf * math.cos(theta_radians) + 
                                                        self.computer.cheap * math.sin(theta_radians)) ** self.computer_market.rho
            
            # 计算计算机生产成本 (方程11)
            self.computer.production_cost = self.computer_market.nu_computer / self.computer.cheap
        
    def accounting(self):
        """
        更新公司的销售和利润账户
        """
        # 方程6
        self.profit = self.q_sold * self.computer.production_cost * self.computer_market.markup
        
    def calc_share(self, q_market):
        """
        计算公司的市场份额
        
        Args:
            q_market: 市场总销售量
        """
        if q_market > 0:
            self.share = self.q_sold / q_market
        else:
            self.share = 0.0
            
    def check_exit(self):
        """
        检查是否仍然满足留在行业的条件
        """
        # 方程20
        self.exit_share = self.computer_market.weight_exit * self.share + (1 - self.computer_market.weight_exit) * self.exit_share
        
        random_number = self.computer_market.rng.random()
        if self.exit_share < self.computer_market.exit_threshold and random_number < 0.5:
            self.exit_firm()
            
    def exit_firm(self):
        """
        退出时激活的方法：将公司活动控制器切换为"False"并重置公司和产品层面的最相关变量
        """
        self.alive = False
        self.born = False
        self.component_rd = 0.0
        self.exit_share = 0.0
        self.num_of_draws_cmp = 0
        self.num_of_draws_sys = 0
        self.price = 0.0
        self.profit = 0.0
        self.prop_to_int = 0.0
        self.prop_to_spec = 0.0
        self.q_sold = 0.0
        self.share = 0.0
        self.system_rd = 0.0
        
        # 如果是专业化公司，通知供应商取消合同
        if not self.integrated and self.supplier_id != -1 and self.computer_market.model:
            component_firm = self.computer_market.model.component.firm[self.supplier_id]
            component_firm.cancel_contract(self.id)
            self.supplier_id = -1
            
        # 重置产品层面变量
        self.computer.cheap = 0.0
        self.computer.mod = 0.0
        self.computer.perf = 0.0
        self.computer.production_cost = 0.0
        self.computer.u = 0.0
        self.computer.U = 0.0
        
        self.system.mod = 0.0
        self.system.mu_prog = 0.0
        
        if self.integrated:
            self.component.mod = 0.0
            self.component.production_cost = 0.0
            self.component.mu_prog = 0.0 

File: src/Chapter4/ComputerFirm.java
package Chapter4;

/*                                                                            *
 * @author Gianluca Capone & Davide Sgobba                                    *
 *                                                                            */

/* This class contains all variables that define firm heterogeneity in the    *
 * computer market (either Mainframes or PC market) and the methods that      *
 * operate over these variables                                               */
class ComputerFirm {
    
    // VARIABLES
    double componentRd;
    /* Amount of resources invested in component R&D (B-CO_f,t)               */
    double exitShare;
    /* Inverse exit propensity (E_f,t)                                        */
    int numOfDrawsCmp;
    /* Number of new potential component mods drawn by the firm (N-CO_f,t)    */
    int numOfDrawsSys;
    /* Number of new potential system mods drawn by the firm (N-SY_f,t)       */
    double probToInt;
    /* Integration probability (I_f,t)                                        */
    double propToInt;
    /* Integration propensity (i_f,t)                                         */
    double probToSpec;
    /* Specialization probability (SIGMA_f,t)                                 */
    double propToSpec;
    /* Specialization propensity (sigma_f,t)                                  */
    double price;
    /* Unit price charged by the computer firm (P_f,t))                       */
    double profit;
    /* Amount of profit earned by the firm (PI_f,t)                           */
    double qSold;
    /* Quantity of computers sold to customers (q_f,t)                        */
    double share;
    /* Market share of the firm (s_f,t)                                       */
    double systemRd;
    /* Amount of resources invested in system R&D (B-SY_f,t)                  */
    
    // PARAMETERS
    double spillover;
    /* Spillover of systemr R&D to component R&D in integrated firms (PSI-CO) */
                                                              
    //TECHNICAL VARIABLES & OBJECTS
    boolean alive;
    /* It takes value "TRUE" if the firm is alive and active on the market,   *
     * "FALSE" otherwise                                                      */
    boolean born;
    /* All firms start with "TRUE"- It takes the value "FALSE" as soon as     *
     * they select the first supplier                                         */
    int contractD;
    /* Duration of the current contract with the supplier                     */
    int contractTime;
    /* Starting period of the current contract with the supplier              */
    int id;
    /* Identifier of the firm                                                 */
    boolean integrated;
    /* It takes value "TRUE" if the firm is vertically integrated, "FALSE"    *
     * otherwise. All firms start with value "FALSE"                          */
    int intTime;
    /* Starting period of integration                                         */
    int supplierId;
    /* Identifier of current supplier                                         */
    int tId;
    /* Identifier of the Component technology. 0 = Transistor technology;     * 
     * 1 = Integrated Circuit technology; 2 = Microprocessor technology       */
    ComputerMarket ComputerMarket;
    /* Access to Computer Market                                              */
    EndProduct Computer;
    /* Computer product to be sold to customers                               */
    NotSoldComponent Component;
    /* Component element to be used in the computer                           */
    SystemElement SysteM;
    /* System element to be used in the computer                              */
    
    // CONSTRUCTOR
    ComputerFirm(int ID, boolean PC, double STARTSHARE, double SPILLOVER,
                 double MODSYS, ComputerMarket COMPUTERMARKET) {
        
        id             = ID;
        exitShare      = STARTSHARE;
        spillover      = SPILLOVER;
        alive          = true;
        born           = true;
        integrated     = false;
        tId            = 0;
        componentRd    = 0;
        profit         = 0;
        intTime        = 0;
        contractTime   = 0;
        contractD      = 0;
        supplierId     = -1;
        
        SysteM         = new SystemElement(MODSYS, this);
        Component      = new NotSoldComponent(this);
        Computer       = new EndProduct(this);
        ComputerMarket = COMPUTERMARKET;
    }

    /* This method determines the probability to integrate if the firm is a   *
     * specialized computer producer                                          */
    void calcProbToIntegrate(double QCO, int T) {
        
        if (QCO > 0) {
            /* Equation 16                                                    */
            propToInt = Math.pow(Math.min((T - ComputerMarket.entryTimeCmpTec[ComputerMarket.tIdCmp]) / ComputerMarket.chi0, 1), ComputerMarket.chi1)
                      * Math.pow((qSold / QCO), ComputerMarket.chi2);
            /* Equation 17                                                    */
            probToInt = (ComputerMarket.xiInt * propToInt)
                      / (1 + (ComputerMarket.xiInt * propToInt));
        } else {
            probToInt = 1;
        }
        
    }

    /* This method implements firm-level changes in case of integration       */
    void integrateVertically(ComponentFirm SUPPLIER, int TIME) {
        integrated            = true;
        Component.mod         = ComputerMarket.inheritance * SUPPLIER.Component.mod;
        tId                   = SUPPLIER.tId;
        intTime               = TIME;
        supplierId            = -1;
        contractTime          = 0;
        contractD             = 0;
    }
    
    /* This method determines the probability to specialize if the firm is an *
     * integrated computer producer                                           */
    void calcProbToSpecialize(double BESTCMPMOD) {
        
        /* Equation 18                                                        */
        propToSpec = (Math.max((BESTCMPMOD - Component.mod) / Component.mod, 0));
        /* Equation 19                                                        */
        probToSpec = (ComputerMarket.xiSpec * propToSpec)
                   / (1 + (ComputerMarket.xiSpec * propToSpec));
        
    }
    
    /* This method implements firm-level changes in case of specialization or *
     * when a new supplier is chosen                                          */
    void signContract(int SUPPLIERID, int TIME, int TID) {
        integrated            = false;
        supplierId            = SUPPLIERID;
        contractTime          = TIME;
        tId                   = TID;
        contractD             = ComputerMarket.minLengthContr
                              + ComputerMarket.rng.nextInt(ComputerMarket.rangeLengthContr);
        intTime               = 0;
    }

    /* This method computes the R&D expenditure and allocates the resources   *
     * to Component R&D and System R&D in the case of integrated firms        */
    void rdExpenditure() {
        
        systemRd = ComputerMarket.rdOnProf * profit;
        if (integrated) {
            /* Equation 12                                                    */
            systemRd -= ComputerMarket.rdOnProf * qSold * ComputerMarket.markup * Component.productionCost;
            if (systemRd > 0) {
                componentRd = (spillover * systemRd
                             + ComputerMarket.rdOnProf * qSold * ComputerMarket.markup * (Component.productionCost));
            } else {
                /* See footnote 15                                            */
                systemRd = 0;
                componentRd = ComputerMarket.rdOnProf * profit;
            }
        }
    }
    
    /* This method checks whether the integrated firm can switch to a new     *
     * component technology because of the high level of the public knowledge */
    void checkComponentTechnology(int TIME, double[] PK) {
        if (tId == 0) {
            if (TIME >= ComputerMarket.entryTimeCmpTec[1]) {
                double meanTemp = Math.log(PK[0]) * (1 - ComputerMarket.internalCum)
                                + Math.log(Component.mod) * ComputerMarket.internalCum;
                if (Math.log(PK[1]) > meanTemp) {
                    tId = 1;
                }
            }
        }
        if (tId == 1) {
            if (TIME >= ComputerMarket.entryTimeCmpTec[2]) {
                double meanTemp = Math.log(PK[1]) * (1 - ComputerMarket.internalCum)
                                + Math.log(Component.mod) * ComputerMarket.internalCum;
                if(Math.log(PK[2]) > meanTemp) {
                    tId = 2;
                }
            }
        }
    }
    
    /* This method determines the outcomes of the innovation activity of the  *
     * firm, determining the level of mod of the component produced by the    *
     * integrated computer firm                                               */
    void progressCmp() {
        
        double tempNumOfDraws = componentRd / ComputerMarket.drawCostCmp[tId];
        numOfDrawsCmp         = (int) Math.floor(tempNumOfDraws);
        double remain         = tempNumOfDraws - numOfDrawsCmp;
        double randomNumber   = ComputerMarket.rng.nextDouble();
        
        if (randomNumber <= remain) {
            numOfDrawsCmp++;
        }

        Component.calcMod();
    }
    
    /* This method determines the outcomes of the innovation activity of the  *
     * firm, determining the level of mod of the system produced by the       *
     * computer firm                                                          */
    void progressSys() {
        
        double tempNumOfDraws = systemRd / ComputerMarket.drawCostSys;
        numOfDrawsSys         = (int) Math.floor(tempNumOfDraws);
        double remain         = tempNumOfDraws - numOfDrawsSys;
        double randomNumber   = ComputerMarket.rng.nextDouble();
        
        if (randomNumber <= remain) {
            numOfDrawsSys++;
        }

        SysteM.calcMod();
    }
    
    /* This method computes the computer mod, and from it derives perfomance, *
     * cheapness, price, and production cost                                  */
    void computerModPriceCost() {

        Computer.calcMod();
        Computer.calcCheapPerf();
        if (Computer.mod > 0) {
            /* Equation 8                                                     */
            price =  ComputerMarket.nuComputer / (Computer.cheap);
        } else {
            price = 0;
        }
        Computer.calcCost();
    }
    
    /* This method computes the sales of the computer firm                    */
    void calcQuantitySold(int NUMOFEXTMKTS) {
        qSold = NUMOFEXTMKTS*Computer.modForCust;
    }
    
    /* This method updates the profits account of the firm                    */
    void accounting() {
        if (!integrated) {
            /* Equation 10                                                    */
            profit     = (Computer.productionCost * ComputerMarket.markup * qSold)
                       - (qSold * ComputerMarket.markup * Component.productionCost);
        }
        else {
            /* Equation 6                                                     */
            profit     = (Computer.productionCost * ComputerMarket.markup * qSold);
        }
    }
    
    /* This method computes the market share of the firm                      */
    void calcShare(double TOTSOLD) {
        if (TOTSOLD != 0) {
            share = qSold / TOTSOLD;
        } else {
            share = 0;
        }
    }
    
    /* This method checks whether the conditions to remain in the industry    *
     * still hold                                                             */
    boolean exitTheMarket() {
        
        boolean exit = false;
        exitShare = ((1 - ComputerMarket.weightExit) * exitShare)
                  + (ComputerMarket.weightExit * share);

        if (exitShare < ComputerMarket.exitThreshold) {
            exit = true;
        }
        return exit;
    }
    
    /* This method is activated when exit occurs: it switches the firm        *
     * controller of activitity to "FALSE" and resets the most relevant       *
     * variables at the firm and product level                                */
    void exitFirm() {
        alive        = false;
        supplierId   = -1;
        contractTime =  0;
        contractD    =  0;
        probToInt    = -1;
        propToInt    = -1;
        probToSpec   = -1;
        propToSpec   = -1;
        profit       =  0;
        systemRd     =  0;
        spillover    =  0;
        componentRd  =  0;
        price        =  0;
        qSold        =  0;
        share        =  0;
        Component.exitComponent();
        SysteM.exitSystem();
        Computer.exitComputer();
    }
    
}
